### DISCUSSION TITLE
Artillery Expect Plugin

### DISCUSSION THREAD

**Discussion by johnnyrockoff** [artilleryio/artillery#1656] [Artillery]
I'm creating a solution to update Test Cases in Azure Test Plans Section and send Adaptive Cards on Azure Devops Teams Channels if a test fail. We are using the expect plugin and when a test fail because of the expect the artillery does not call the afterResponse section on my yaml file, is there any way to use expect without breaking the flow?

**Comment by juanig1** [comment]
Hey!
Correct me if I am wrong but: you are using the `afterResponse` plus some custom logic flow to dispatch notifications if some conditions fail.

One alternative to that could be to use `ensure` to force those conditions and then, if the run fails, use the exit code of the artillery command (it will be non-zero if test run failed because of ensure) and handle the notification flow there.

Super simple example:
```artillery run my-script-with-ensures.yaml || echo "something went wrong!"```
Will print `something went wrong!` if test execution fails

**Reply by meetveera** [reply]
But it will not solve the problem to send failed responses to Azure Test Pans right? 
Here if any API reposne is other than expected than the afterResponse hook is not executed thus resulting to not  been abe to send failed API data to Azure Test Plan

**Comment by johnnyrockoff** [comment]
@juanig1 There is another point in this case, I want to get the response from the flows like statusCode and statusMessage and send it through a custom javascript function.

" artillery run my-script-with-ensures.yaml || echo "something went wrong!" " using this solution the script is not dynamically loaded with the response.

**Comment by juanig1** [comment]
Hey, thanks for the clarification, wasn't clear for me you used request data in your notification flow.

There is no explicit option to execute `afterReponse` after ensure fails.
But: maybe setting [`strict:false`](https://www.artillery.io/docs/guides/guides/test-script-reference#advanced-conditional-checks) and letting the test continue is enough for your case (you may need to add extra logic to your custom js function to handle test halting).
I don't now how your scripts looks, but another approach could be using [custom metrics](https://www.artillery.io/docs/guides/guides/extension-apis#tracking-custom-metrics) and using `ensure` to check on those (for example, on a failed request: handle notification and then emitting to trigger the `ensure` logic)

**Reply by johnnyrockoff** [reply]
I'll try! Thanks! I'll give you a feedback!

**Reply by johnnyrockoff** [reply]
I had no success! :(
